### ISSUE TITLE
Framework error codes should provide more informative diagnostics on error

### ISSUE THREAD

**Issue by armaniferrante** [project-serum/anchor#823]
Instead of returning error codes, the framework should either panic on error or embed the line number and filename into the error type to provide more informative diagnostic information.

Potential resources

* https://doc.rust-lang.org/std/panic/fn.set_hook.html
* https://crates.io/crates/panic-context

**Comment by armaniferrante** [comment]
Blocked by https://github.com/project-serum/anchor/issues/798.

**Comment by armaniferrante** [comment]
Closing in favor of https://github.com/project-serum/anchor/issues/798.
